### PR TITLE
Add some typings informations and typecheck

### PR DESCRIPTION
**Pull request by Carreau**
Mostly make it a bit clearer what is allowed/expected where.
In particular many of the decode fucntions allow already decoded
objects.

TODO:
* [x] Add unit tests and/or doctests in docstrings
* [x] Add docstrings and API docs for any new/modified user-facing classes and functions
* [x] New/modified features documented in docs/tutorial.rst
* [x] Changes documented in docs/release.rst
* [x] AppVeyor and Travis CI passes
* [x] Test coverage is 100% (Coveralls passes)

### CODE REVIEW

**Comment by Carreau** [comment]
Added  Typing to storage.py.

Note in particular that some function can get bytes/str/anything that
can be stringified though others only can get str/None as they don't call normalize on path.

**Review by alimanfoo — APPROVED** [review]

**Review by jakirkham — APPROVED** [review]

**Comment by Carreau** [comment]
rebased to fix conflicts in `pytest.ini`, no other changes.

**Comment by Carreau** [comment]
All tests are passing and the push was only to fix a trivial conflict in pytest.ini merging. Thanks.

File: zarr/meta.py [hunk 1]
```diff
@@ -74,27 +76,27 @@ def encode_array_metadata(meta):
     return json_dumps(meta)
 
 
-def encode_dtype(d):
+def encode_dtype(d: np.dtype) -> str:
```
Comment (joshmoore) [inline zarr/meta.py:79]: see https://github.com/zarr-developers/zarr-python/pull/700

**Pull request merged**